proxy/src/control/discovery.rs

```diff
@@ -58,6 +63,7 @@ pub struct DiscoveryWork<T: HttpService<ResponseBody = RecvBody>> {
 struct DestinationSet<T: HttpService<ResponseBody = RecvBody>> {
     addrs: Exists<Cache<SocketAddr>>,
     query: Option<DestinationServiceQuery<T>>,
+    dns_query: Option<IpAddrListFuture>,
```

Comment:
just to confirm my understanding...  it's the case that when we have a `dns_query`, we're guaranteed to also have a `query`?  Does it make sense to have a single `Query<T>`  type that holds both the `DestinationServiceQuery<T>` and `Option<IpAddrListFuture>`? tioli... just asking questions as i read.

Reply:
> It's the case that when we have a dns_query, we're guaranteed to also have a query?

No. When we first create the `DestinationSet`, we check if the query is `None` and then automatically create the `dns_query` if it is None. So all combinations are possible.

Reply:
Hmm. Are there specific states in which we would expect a given combination of `None` and `Some`? If so we might want to consider an `enum` that encodes those states and the allowable transitions between them...

Reply:
`(None, None)` is not a valid state but intermittently we do have `(None, None)` sometimes because of the uses of `take` in `poll_destinations()`.

Reply:
As far as the transitions go:

`(Some, None)` is a valid initial state, when the name is of a form for which we'd query the Destination service.
`(None, Some)` is a valid initial state, when that's not the case.

`(Some, None) => (Some, Some)` when the destination service indicates that the service doesn't exist.
`(Some, Some) => (Some, None)` when the destination service indicates the service exists after previously telling us it didn't.

`(None, None)` is a frequent intermediate state but it isn't a valid state in general.

We could try to encode this into a typed state machine, but I think that would impede immediately-upcoming work to generalize service discovery beyond the Destination service and DNS. I suggest we add that 3rd type of service discovery and do that refactoring before we fossilize the current design too much more. In particular, in the future it would be nice to not have `control::discovery` aware of the Destination service or DNS at all. But not yet, and not in this PR. :)

Reply:
👍

Reply:
Okay, yeah, that seems entirely reasonable to me --- I was just curious. 👍